### PR TITLE
Json logging

### PR DESCRIPTION
**Pull request by davilimap**
Adding logging capabilities for JSON transformations

### CODE REVIEW

File: SlowCheetah.JDT/JdtException.cs [hunk 1]
```diff
@@ -1,6 +1,31 @@
...
+
+namespace SlowCheetah.JDT
 {
     using System;
+    using Newtonsoft.Json;
+
+    /// <summary>
+    /// The file that caused the exception
+    /// </summary>
+    internal enum ErrorLocation
```
Comment (jviau) [inline SlowCheetah.JDT/JdtException.cs]: I don't think `internal` is necessary here. Could be `public`. Along with all the non-newtonsoft contructors and methods in `JdtException`

File: SlowCheetah.JDT/JsonTransformation.cs [hunk 2]
```diff
@@ -13,17 +16,31 @@ public class JsonTransformation
         private JObject transformObject;
         private JsonLoadSettings loadSettings;
 
+        private JsonTransformationContextLogger logger;
```
Comment (jviau) [inline SlowCheetah.JDT/JsonTransformation.cs]: Can this, or any other fields, be `readonly`?

File: SlowCheetah.JDT/JsonTransformation.cs [hunk 3]
```diff
@@ -35,39 +52,54 @@ public JsonTransformation(string transformFile)
...
 
         /// <summary>
         /// Transforms a JSON object
         /// </summary>
         /// <param name="sourceFile">The path to the json file to be transformed</param>
-        /// <returns>A stream containing the results of the transforms</returns>
-        public Stream Apply(string sourceFile)
+        /// <param name="result">The stream to write the result into</param>
+        /// <returns>True if the transformations were completed</returns>
+        public bool Apply(string sourceFile, out Stream result)
```
Comment (jviau) [inline SlowCheetah.JDT/JsonTransformation.cs]: Rename to `TryApply` (and rename the one below also) and add back the previous `Apply` methods where it throws exceptions if it fails.

File: SlowCheetah.JDT/JsonTransformation.cs [hunk 4]
```diff
@@ -35,39 +52,54 @@ public JsonTransformation(string transformFile)
...
             if (string.IsNullOrEmpty(sourceFile))
             {
                 throw new ArgumentNullException(nameof(sourceFile));
             }
 
+            this.logger.SourceFile = sourceFile;
+
             // Open the file as streams and apply the transforms
-            // TO DO: Save the source and transform file to the context for logging
             Stream sourceStream = File.Open(sourceFile, FileMode.Open);
```
Comment (jviau) [inline SlowCheetah.JDT/JsonTransformation.cs]: This needs to be in a `using`, you are not disposing the stream.

File: SlowCheetah.JDT/JsonTransformation.cs [hunk 5]
```diff
@@ -86,26 +118,44 @@ public Stream Apply(Stream source)
...
+                    // The JObject corresponding to the streams with line info
+                    sourceObject = JObject.Load(sourceReader, loadSettings);
 
-                return result;
+                    // Execute the transforms
+                    JdtProcessor.ProcessTransform(sourceObject, this.transformObject, this.logger);
+                }
+                catch (Exception ex)
+                {
+                    if (!this.logger.LogErrorFromException(ex))
```
Comment (jviau) [inline SlowCheetah.JDT/JsonTransformation.cs]: The `TryApply` version should not throw for a failed transform. It should only fail for things like file load failure.

File: SlowCheetah.JDT/JsonTransformationContextLogger.cs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+{
+    using System;
+    using Newtonsoft.Json;
+
+    /// <summary>
+    /// Logger wrapper for JDT transformations
+    /// </summary>
+    internal class JsonTransformationContextLogger
+    {
+        private IJsonTransformationLogger externalLogger = null;
```
Comment (jviau) [inline SlowCheetah.JDT/JsonTransformationContextLogger.cs]: `readonly`

File: SlowCheetah.JDT/JsonTransformationContextLogger.cs [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        internal JsonTransformationContextLogger(string transformationFile, IJsonTransformationLogger extLogger)
+        {
+            this.externalLogger = extLogger;
+            this.TransformFile = transformationFile;
+        }
+
+        /// <summary>
+        /// Gets or sets the source file of the current transformation
+        /// </summary>
+        internal string SourceFile { get; set; } = "Source";
```
Comment (jviau) [inline SlowCheetah.JDT/JsonTransformationContextLogger.cs]: Any reason for the `set`? It is not used. Same with `TransformFile`.

File: SlowCheetah.JDT/JsonTransformationContextLogger.cs [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+        /// <summary>
+        /// Gets or sets the transformation file of the current transformation
+        /// </summary>
+        internal string TransformFile { get; set; } = "Transform";
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the logger has logged errrors
+        /// </summary>
+        internal bool HasLoggedErrors { get; set; }
```
Comment (jviau) [inline SlowCheetah.JDT/JsonTransformationContextLogger.cs]: `private set`

File: SlowCheetah.JDT/JsonTransformationContextLogger.cs [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        private string LocationPath(ErrorLocation location)
+        {
+            switch (location)
+            {
+                case ErrorLocation.Source:
+                    return this.SourceFile;
+                case ErrorLocation.Transform:
+                    return this.TransformFile;
+                default:
+                    return null;
```
Comment (jviau) [inline SlowCheetah.JDT/JsonTransformationContextLogger.cs:148]: What happens if this does end up returning `null`? Looking at your uses of it, you pass the result directly into `externalLogger`. What if it throws for `null`?

File: SlowCheetah.JDT/JsonTransformationContextLogger.cs [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+                    }
+                }
+
+                // The exception has been logged
+                return true;
+            }
+            else
+            {
+                // The exception has not been logged
+                return false;
```
Comment (jviau) [inline SlowCheetah.JDT/JsonTransformationContextLogger.cs]: As discussed in person, rethrow the exception here using `ExceptionDispathInfo.Capture(exception).Throw();` Then switch to a `void` return.

**Review by jviau — CHANGES_REQUESTED** [review]
1. You added the `logger` parameter everywhere - add argument validation for it on all non-private methods.
2. Look over the new types added and add `readonly` and remove `set` (or switch to `private set`) where possible.

File: SlowCheetah.JDT/JsonTransformationContextLogger.cs (hunk 9 above)
Comment (davilimap) [reply SlowCheetah.JDT/JsonTransformationContextLogger.cs:148]: I think the `IJsonTransformationLogger` should be able to handle null values for the file path. It's in the XML doc for the interface.

**Comment by davilimap** [comment]
@jviau Addressed all of your suggestions and made a few more changes

File: SlowCheetah.JDT/JsonTransformation.cs [hunk 11]
```diff
@@ -79,7 +79,7 @@ public JsonTransformation(Stream transform, IJsonTransformationLogger logger)
         /// <param name="sourceFile">The path to the json file to be transformed</param>
         /// <param name="result">The stream to write the result into</param>
         /// <returns>True if the transformations were completed</returns>
-        public bool Apply(string sourceFile, out Stream result)
+        public bool TryApply(string sourceFile, out Stream result)
```
Comment (jviau) [inline SlowCheetah.JDT/JsonTransformation.cs]: We need to rethink this API. I didn't realize in my previous comment the logger was part of the constructor. `TryApply` will currently throw when the user has no logger supplied, which is not what a `Try` function should do. I think we need an `Apply` overload that propagates the exception regardless of logger being provided. Then the `TryApply` never throws logger or not.

**Review by jviau — CHANGES_REQUESTED** [review]

File: SlowCheetah.JDT/JsonTransformation.cs (hunk 11 above)
Comment (jviau) [reply SlowCheetah.JDT/JsonTransformation.cs]: Other than this issue it looks good.

**Comment by davilimap** [comment]
@jviau Performed changes in the entry points as discussed.

**Comment by jviau** [comment]
Sorry to go back and forth on the design of TryApply vs Apply, but as discussed in person lets remove the TryApply for now. It hides too much information of why it failed to the caller. We can revisit a TryApply method in the future if the demand is there.

**Review by jviau — APPROVED** [review]

**Pull request merged**